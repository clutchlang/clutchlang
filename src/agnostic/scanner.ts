import { splitLines } from './strings';

/**
 * Represents a section of @member text.
 *
 * Contains enough information in order to represent actionable error or
 * informational messages. Some members in some implementations may be lazy.
 */
export class SourceSpan {
  private static readonly matchNewLine = /\n|\r/;

  constructor(
    public readonly offset: number,
    public readonly column: number,
    public readonly line: number,
    public readonly text: string
  ) {
    if (offset < 0) {
      throw new RangeError(`Invalid offset: ${offset}`);
    }
    if (column < 0) {
      throw new RangeError(`Invalid column: ${column}`);
    }
    if (line < 0) {
      throw new RangeError(`Invalid line: ${line}`);
    }
  }

  /**
   * Length of the span of information.
   */
  public get length(): number {
    return this.text.length;
  }

  /**
   * All lines, separated by line number and text, present in @member text.
   *
   * **NOTE**: `\n\r` is interpreted as a single line break, not two.
   */
  public get lines(): Array<{ line: number; text: string }> {
    return this.isMultiLine
      ? this.computeMultiLines()
      : [
          {
            line: this.line,
            text: this.text,
          },
        ];
  }

  /**
   * Whether the @member text spans over multiple lines.
   *
   * When true, @member lines will represent multiple line numbers.
   */
  public get isMultiLine(): boolean {
    return SourceSpan.matchNewLine.test(this.text);
  }

  private computeMultiLines(): Array<{ line: number; text: string }> {
    return Array.from(
      splitLines(this.text).map((text, index) => {
        return {
          line: this.line + index,
          text,
        };
      })
    );
  }
}

/**
 * A function that handles receiving a @param span and @param error.
 *
 * May optionally return token(s) to recover from the error. Otherwise it is
 * expected the error reporting was fatal and that parsing should be immediately
 * stopped.
 */
export type ErrorReporter = (
  span: SourceSpan,
  error: ScanningError
) => Token | Token[] | void;

/**
 * Error types that may be provided to a @see ErrorReporter.
 */
export type ScanningError = UnexpectedTokenError;

/**
 * A syntax error that occurs as a result of an unexpected @member token.
 *
 * Optionally contains a @member context, or what was being parsed when the
 * error ocurrred, i.e. "argument list" or "function body". May be omitted if
 * the context is not clear or provided.
 */
export class UnexpectedTokenError extends SyntaxError {
  constructor(public readonly token: Token, public readonly context?: string) {
    super(`Unexpected "token"${context ? ` in ${context}` : ''}.`);
  }
}

/**
 * A token that was scanned.
 */
export class Token {}
