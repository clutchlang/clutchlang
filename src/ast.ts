export * from './language/ast/token';
