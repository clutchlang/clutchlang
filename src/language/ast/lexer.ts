export * from './lexer/token';
export * from './lexer/tokenizer';
