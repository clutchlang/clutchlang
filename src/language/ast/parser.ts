import * as tokens from './token';

/**
 * Represents an unresolved node in the abstract syntax tree.
 */
export abstract class AstNode {
  public abstract accept<R, C>(
    visitor: AstVisitor<R, C>,
    context?: C
  ): R | undefined;

  /**
   * First token that represents this AST node.
   */
  public abstract get first(): tokens.Token;

  /**
   * Last token that represents this AST node.
   */
  public abstract get last(): tokens.Token;
}

/**
 * Represents a simple AST node that originates from a single token.
 */
export abstract class SimpleNode extends AstNode {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an operator node parsed from a single token.
 */
export class Operator extends SimpleNode {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitOperator(this, context);
  }
}

/**
 * Represents a marker type of AST node for all statements.
 */
export abstract class Statement extends AstNode {}

/**
 * Represents a marker type of AST node for all expressions.
 */
export abstract class Expression extends Statement {}

/**
 * Represents a simple AST node that is also an expression.
 */
export abstract class SimpleExpression extends Expression {
  constructor(protected readonly token: tokens.Token) {
    super();
  }

  public get first(): tokens.Token {
    return this.token;
  }

  public get last(): tokens.Token {
    return this.token;
  }
}

/**
 * Represents an expression that utilizes an operator.
 */
export abstract class OperatorExpression extends Expression {
  constructor(
    public readonly operator: Operator,
    public readonly target: Expression
  ) {
    super();
  }
}

/**
 * Represents an expression in the form of `<OP><EXPR>`.
 */
export class PrefixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPrefixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.operator.first;
  }

  public get last(): tokens.Token {
    return this.target.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP>`.
 */
export class PostfixExpression extends OperatorExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPostfixExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.operator.last;
  }
}

/**
 * Represents an expression in the form of `<EXPR><OP><EXPR>`.
 */
export class BinaryExpression extends OperatorExpression {
  constructor(
    operator: Operator,
    public readonly left: Expression,
    public readonly right: Expression
  ) {
    super(operator, left);
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitBinaryExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.left.first;
  }

  public get last(): tokens.Token {
    return this.right.last;
  }
}

/**
 * Represents an expression in the form of `<TARGET>.<PROPERTY>`.
 */
export class PropertyExpression extends Expression {
  constructor(
    public readonly target: Expression,
    public readonly property: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitPropertyExpression(this, context);
  }

  public get first(): tokens.Token {
    return this.target.first;
  }

  public get last(): tokens.Token {
    return this.property.last;
  }
}

/**
 * Represents a list of arguments passed to a function call.
 */
export class ArgumentList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly args: Expression[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitArgumentList(this, context);
  }
}

/**
 * Represents an expression for a function call.
 *
 * May _optionally_ contain a @member target.
 */
export class CallExpression extends Expression {
  constructor(
    public readonly name: Identifier,
    public readonly args: ArgumentList,
    public readonly target?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitCallExpression(this, context);
  }

  public get first(): tokens.Token {
    return (this.target || this.name).first;
  }

  public get last(): tokens.Token {
    return this.args.last;
  }
}

export class ConditionalExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly condition: Expression,
    public readonly body: Expression | StatementBlock,
    public readonly elseCondition?: Expression,
    public readonly elseBody?: Expression | StatementBlock
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitConditionalExpression(this, context);
  }

  public get last(): tokens.Token {
    return (this.elseBody || this.body).last;
  }
}

export class GroupExpression extends Expression {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression: Expression,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitGroupExpression(this, context);
  }
}

/**
 * Represents an expression for a literal identifier (name).
 */
export class Identifier extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitIdentifier(this, context);
  }

  public get name(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal boolean.
 */
export class LiteralBoolean extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralBoolean(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal number.
 */
export class LiteralNumber extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralNumber(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents an expression for a literal string.
 */
export class LiteralString extends SimpleExpression {
  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitLiteralString(this, context);
  }

  public get value(): string {
    return this.token.lexeme;
  }
}

/**
 * Represents a variable declared anywhere in scope.
 */
export class VariableDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly type?: Identifier,
    public readonly value?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitVariableDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return (this.value || this.type || this.name).last;
  }
}

/**
 * Represents a list of parameters for a function or closure definition.
 */
export class ParameterList extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly params: VariableDeclaration[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitParameterList(this, context);
  }
}

/**
 * Represents a series of statements.
 */
export class StatementBlock extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly statements: Statement[],
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitStatementBlock(this, context);
  }
}

/**
 * Represents a return statement.
 */
export class ReturnStatement extends AstNode {
  constructor(
    public readonly first: tokens.Token,
    public readonly expression?: Expression
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitReturnStatement(this, context);
  }

  public get last(): tokens.Token {
    return this.expression ? this.expression.last : this.first;
  }
}

/**
 * Represents a function declared anywhere in scope.
 */
export class FunctionDeclaration extends AstNode {
  constructor(
    public readonly name: Identifier,
    public readonly params: ParameterList,
    public readonly body: Expression | StatementBlock,
    public readonly returnType?: Identifier
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitFunctionDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.name.first;
  }

  public get last(): tokens.Token {
    return this.body.last;
  }
}

/**
 * Represents a type declared anywhere in scope.
 */
export class TypeDeclaration extends AstNode {
  constructor(
    public readonly modifiers: tokens.Token[],
    public readonly name: Identifier,
    public readonly members: Array<FunctionDeclaration | VariableDeclaration>,
    public readonly last: tokens.Token
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitTypeDeclaration(this, context);
  }

  public get first(): tokens.Token {
    return this.modifiers.length ? this.modifiers[0] : this.name.first;
  }
}

/**
 *
 */
export class ModuleRoot extends AstNode {
  constructor(
    public readonly declarations: Array<
      FunctionDeclaration | TypeDeclaration | VariableDeclaration
    >
  ) {
    super();
  }

  public accept<R, C>(visitor: AstVisitor<R, C>, context?: C): R | undefined {
    return visitor.visitModuleRoot(this, context);
  }

  public get first(): tokens.Token {
    return this.declarations[0].first;
  }

  public get last(): tokens.Token {
    return this.declarations[this.declarations.length - 1].last;
  }
}

export class AstParser {}

/**
 * An AST visitor that by default does nothing visiting an AST node.
 *
 * It is intended to be a superclass for classes that use the visitor pattern
 * primarily as a dispatch mechanism (and do not need to recursively need to
 * visit the whole structure) and only need to visit a small number of node
 * types.
 */
export class AstVisitor<R, C> {
  public visitArgumentList(_: ArgumentList, __?: C): R | undefined {
    return;
  }

  public visitBinaryExpression(_: BinaryExpression, __?: C): R | undefined {
    return;
  }

  public visitCallExpression(_: CallExpression, __?: C): R | undefined {
    return;
  }

  public visitConditionalExpression(
    _: ConditionalExpression,
    __?: C
  ): R | undefined {
    return;
  }

  public visitFunctionDeclaration(
    _: FunctionDeclaration,
    __?: C
  ): R | undefined {
    return;
  }

  public visitGroupExpression(_: GroupExpression, __?: C): R | undefined {
    return;
  }

  public visitIdentifier(_: Identifier, __?: C): R | undefined {
    return;
  }

  public visitLiteralBoolean(_: LiteralBoolean, __?: C): R | undefined {
    return;
  }

  public visitLiteralNumber(_: LiteralNumber, __?: C): R | undefined {
    return;
  }

  public visitLiteralString(_: LiteralString, __?: C): R | undefined {
    return;
  }

  public visitModuleRoot(_: ModuleRoot, __?: C): R | undefined {
    return;
  }

  public visitOperator(_: Operator, __?: C): R | undefined {
    return;
  }

  public visitParameterList(_: ParameterList, __?: C): R | undefined {
    return;
  }

  public visitPrefixExpression(_: PrefixExpression, __?: C): R | undefined {
    return;
  }

  public visitPostfixExpression(_: PostfixExpression, __?: C): R | undefined {
    return;
  }

  public visitPropertyExpression(_: PropertyExpression, __?: C): R | undefined {
    return;
  }

  public visitReturnStatement(_: ReturnStatement, __?: C): R | undefined {
    return;
  }

  public visitStatementBlock(_: StatementBlock, __?: C): R | undefined {
    return;
  }

  public visitTypeDeclaration(_: TypeDeclaration, __?: C): R | undefined {
    return;
  }

  public visitVariableDeclaration(
    _: VariableDeclaration,
    __?: C
  ): R | undefined {
    return;
  }
}
