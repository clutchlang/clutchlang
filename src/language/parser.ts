export * from './parser/expressions';
export * from './parser/factory';
export * from './parser/nodes';
export * from './parser/operators';
export * from './parser/statements';
export * from './parser/visitors';
