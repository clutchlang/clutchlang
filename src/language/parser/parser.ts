import { FileParser } from './parse/file';

export class ClutchParser extends FileParser {}
