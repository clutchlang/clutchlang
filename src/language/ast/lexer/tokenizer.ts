import { StringScanner } from '../../../agnostic/scanner';
import {
  Characters,
  isDigit,
  isHexadecimal,
  isLetter,
  isWhiteSpace,
} from '../../../agnostic/strings';
import * as tokens from './token';

function isIdentifier(character: number): boolean {
  return isIdentifierStart(character) || isDigit(character);
}

function isIdentifierStart(character: number): boolean {
  return isLetter(character) || character === Characters.$_;
}

/**
 * Tokenizes and returns @param program as a series of @see Token.
 */
export function tokenize(
  program: string,
  onError?: InvalidTokenReporter
): tokens.Token[] {
  return new ClutchLexer().scanProgram(new StringScanner(program), onError);
}

/**
 * Reports a @param message at the provided @param span.
 */
export type InvalidTokenReporter = (offset: number, length: number) => void;

/**
 * A language-level lexer that yields tokens understood by the grammar.
 */
export class ClutchLexer {
  public static readonly keywords: {
    [index: string]: tokens.IKeywordTokenType;
  } = {
    const: tokens.$Const,
    else: tokens.$Else,
    external: tokens.$External,
    false: tokens.$False,
    if: tokens.$If,
    let: tokens.$Let,
    return: tokens.$Return,
    then: tokens.$Then,
    true: tokens.$True,
    type: tokens.$Type,
  };

  private program!: StringScanner;
  private onError!: InvalidTokenReporter;
  private position!: number;
  private lastComments: tokens.Token[] = [];

  public scanProgram(
    program: StringScanner,
    onError: InvalidTokenReporter = (offset, length) => {
      const span = this.program.source.span(offset, offset + length);
      throw new SyntaxError(
        `Invalid token "${span.text}" at ${span.line}:${span.column}`
      );
    }
  ): tokens.Token[] {
    this.position = 0;
    this.program = program;
    this.onError = onError;
    const scanned: tokens.Token[] = [];
    while (program.hasNext()) {
      const token = this.scanToken();
      if (token) {
        scanned.push(token);
      }
    }
    scanned.push(this.createToken(tokens.$EOF, ''));
    return scanned;
  }

  private scanToken(): tokens.Token | undefined {
    const character = this.program.advance();
    switch (character) {
      case Characters.$LPAREN: // "("
        return this.createToken(tokens.$LeftParen);
      case Characters.$RPAREN: // ")"
        return this.createToken(tokens.$RightParen);
      case Characters.$LCURLY: // "{"
        return this.createToken(tokens.$LeftCurly);
      case Characters.$RCURLY: // "}"
        return this.createToken(tokens.$RightCurly);
      case Characters.$COLON: // ":"
        return this.createToken(tokens.$Colon);
      case Characters.$PERIOD: // "."
        return this.createToken(tokens.$Period);
      case Characters.$PLUS: // "+" or "+=" or "++"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$PlusEquals
            : this.program.match(Characters.$PLUS)
              ? tokens.$PlusPlus
              : tokens.$Plus
        );
      case Characters.$MINUS: // "-" or "--" or "-=" or "->"
        return this.createToken(
          this.program.match(Characters.$RANGLE)
            ? tokens.$DashRightAngle
            : this.program.match(Characters.$EQUALS)
              ? tokens.$DashEquals
              : this.program.match(Characters.$MINUS)
                ? tokens.$DashDash
                : tokens.$Dash
          // Code coverage considers the `);` to be a missed line :(
          /* istanbul ignore next */
        );
      case Characters.$STAR: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$StarEquals
            : tokens.$Star
        );
      case Characters.$PERCENT: // "*" or "*="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$PercentEquals
            : tokens.$Percent
        );
      case Characters.$EQUALS: // "=" or "==" or "==="
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? tokens.$EqualsEqualsEquals
              : tokens.$EqualsEquals
            : tokens.$Equals
        );
      case Characters.$LANGLE: // "<" or "<=" or "<<"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$LeftAngleEquals
            : this.program.match(Characters.$LANGLE)
              ? tokens.$LeftAngleLeftAngle
              : tokens.$LeftAngle
        );
      case Characters.$RANGLE: // ">" or ">=" or ">>"
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? tokens.$RightAngleEquals
            : this.program.match(Characters.$RANGLE)
              ? tokens.$RightAngleRightAngle
              : tokens.$RightAngle
        );
      case Characters.$SLASH: // "/" or "//" or "/="
        return this.scanSlash();
      case Characters.$SQUOTE: // "'"
        return this.scanString();
      case Characters.$PIPE: // "|" or "||"
        return this.createToken(
          this.program.match(Characters.$PIPE) ? tokens.$PipePipe : tokens.$Pipe
        );
      case Characters.$AND: // "&" or "&&"
        return this.createToken(
          this.program.match(Characters.$AND) ? tokens.$AndAnd : tokens.$And
        );
      case Characters.$TILDE: // "~"
        return this.createToken(tokens.$Tilde);
      case Characters.$CARET: // "^"
        return this.createToken(tokens.$Caret);
      case Characters.$EXCLAIM: // "!" or "!=" or "!==""
        return this.createToken(
          this.program.match(Characters.$EQUALS)
            ? this.program.match(Characters.$EQUALS)
              ? tokens.$ExclaimEqualsEquals
              : tokens.$ExclaimEquals
            : tokens.$Exclaim
        );
      default:
        if (isWhiteSpace(character)) {
          this.position = this.program.position;
          break;
        }
        if (isDigit(character)) {
          return this.scanNumber(character);
        }
        if (isIdentifierStart(character)) {
          return this.scanIdentifierOrKeyword();
        }
        this.reportError();
    }
    return;
  }

  /**
   * Scans a string, returning `undefined` if there was no string.
   */
  private scanString(): tokens.Token {
    while (this.program.hasNext()) {
      if (this.program.match(Characters.$SQUOTE)) {
        this.position++;
        let lexeme = this.substring();
        lexeme = lexeme.substring(0, lexeme.length - 1);
        return this.createToken(tokens.$String, lexeme);
      }
      this.program.advance();
    }
    this.reportError();
    this.position++;
    return this.createToken(tokens.$String);
  }

  /**
   * Scans any tokens that are valid numeric literals
   */
  private scanNumber(starting: number): tokens.Token {
    if (starting === Characters.$0) {
      if (
        this.program.match(Characters.$X) ||
        this.program.match(Characters.$x)
      ) {
        return this.scanHexNumber();
      }
    }
    if (
      this.program.match(Characters.$E) ||
      this.program.match(Characters.$e)
    ) {
      return this.scanDigits();
    }
    while (this.program.hasNext()) {
      const next = this.program.peek();
      if (this.program.match(Characters.$PERIOD)) {
        break;
      }
      if (!isDigit(next)) {
        break;
      }
      this.program.advance();
    }
    return this.scanDigits();
  }

  /**
   * Scans any tokens that are valid hexadecimel digits, returning a number.
   */
  private scanHexNumber(): tokens.Token {
    this.scanPredicate(isHexadecimal);
    return this.createToken(tokens.$Number);
  }

  /**
   * Scans any tokens that are valid digits, returning a number.
   */
  private scanDigits(): tokens.Token {
    this.scanPredicate(isDigit);
    return this.createToken(tokens.$Number);
  }

  /**
   * Scans any tokens that are valid identifiers.
   */
  private scanIdentifierOrKeyword(): tokens.Token {
    this.scanPredicate(isIdentifier);
    const identifierOrKeyword = this.substring();
    return new tokens.Token(
      this.position - identifierOrKeyword.length,
      ClutchLexer.keywords[identifierOrKeyword] || tokens.$Identifier,
      this.clearComments(),
      identifierOrKeyword
    );
  }

  /**
   * Scans tokens preceding with "/".
   */
  private scanSlash(): tokens.Token | undefined {
    // "//"
    if (this.program.match(Characters.$SLASH)) {
      this.advanceToEOL();
      this.lastComments.push(
        new tokens.Token(
          this.position,
          tokens.$Comment,
          [],
          this.substring().trim()
        )
      );
      return;
    }
    // "/" or "/="
    return this.createToken(
      this.program.match(Characters.$EQUALS)
        ? tokens.$SlashEquals
        : tokens.$Slash
    );
  }

  /**
   * Scans until the EOF or until @param predicate returns false.
   */
  private scanPredicate(predicate: (character: number) => boolean): void {
    while (this.program.hasNext()) {
      const peek = this.program.peek();
      if (!predicate(peek)) {
        break;
      }
      this.program.advance();
    }
  }

  /**
   * Advances the position of the scanner until either EOF or EOL.
   *
   * **NOTE**: Treats `CR+LF` as a DOS-stlye line ending, not two lines.
   */
  private advanceToEOL(): void {
    while (this.program.hasNext()) {
      // "\n" (LF)
      if (this.program.match(Characters.$LF)) {
        return;
      }
      // "\r" (CR)
      if (this.program.match(Characters.$CR)) {
        // "\r\n" (CR+LF)
        this.program.match(Characters.$LF);
        return;
      }
      this.program.advance();
    }
  }

  /**
   * Returns a new token with a @param kind and the current substring.
   */
  private createToken(
    kind: tokens.ITokenTypes,
    content = this.substring()
  ): tokens.Token {
    return new tokens.Token(
      this.position - content.length,
      kind,
      this.clearComments(),
      content
    );
  }

  /**
   * Clear and retrieve comments lexed.
   */
  private clearComments(): tokens.Token[] {
    const comments = this.lastComments;
    this.lastComments = [];
    return comments;
  }

  /**
   * Returns the current substring, advancing the position counter.
   */
  private substring(): string {
    const start = this.position;
    const end = this.program.position;
    this.position = end;
    return this.program.substring(start, end);
  }

  /**
   * Reports a lexical @param error occurring at @param offset.
   */
  private reportError(offset = this.position, length = 1): void {
    this.onError(offset, length);
  }
}
