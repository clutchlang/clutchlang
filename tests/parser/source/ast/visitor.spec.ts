import { AstCompilationUnit } from '../../../../src/parser/source/ast/node';
import { AstParser } from '../../../../src/parser/source/ast/parser';
import { TokenScanner } from '../../../../src/parser/source/ast/scanner';
import { PrintTreeVisitor } from '../../../../src/parser/source/ast/visitor';
import { Lexer } from '../../../../src/parser/source/tokenizer/lexer';
import { SourceScanner } from '../../../../src/parser/source/tokenizer/scanner';

function parse(program: string): AstCompilationUnit {
  const scanner = new SourceScanner(program);
  const lexer = new Lexer(scanner);
  const tokens = Array.from(lexer);
  const parser = new AstParser(new TokenScanner(tokens));
  return parser.parseCompilationUnit();
}

describe(`${PrintTreeVisitor}`, () => {
  it('should emit a readable structure for a program', () => {
    const visitor = new PrintTreeVisitor();
    expect(
      visitor.visitCompilationUnit(
        parse(`
      emptyFunction => {}
      returnsLiteral => true
      hasExpressionBody => {
        1
        2
        3
      }
      hasInvocationBody => {
        print('Hello World')
      }
    `)
      )
    ).toMatchSnapshot();
  });
});
