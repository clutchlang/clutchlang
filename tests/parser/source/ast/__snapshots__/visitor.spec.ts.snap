// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class PrintTreeVisitor extends /* istanbul ignore next */(AstVisitor) {
    constructor(indentBy = /* istanbul ignore next */(cov_1820g0n6e7.b[0][0]++, 2)) {
        /* istanbul ignore next */cov_1820g0n6e7.f[3]++;
        cov_1820g0n6e7.s[8]++;

        super();
        /* istanbul ignore next */cov_1820g0n6e7.s[9]++;
        this.indentBy = indentBy;
        /* istanbul ignore next */cov_1820g0n6e7.s[10]++;
        this.output = '';
        /* istanbul ignore next */cov_1820g0n6e7.s[11]++;
        this.indent = 0;
    }
    visitCompilationUnit(node) {
        /* istanbul ignore next */cov_1820g0n6e7.f[4]++;
        cov_1820g0n6e7.s[12]++;

        this.write('CompilationUnit');
        /* istanbul ignore next */cov_1820g0n6e7.s[13]++;
        this.indentMore();
        /* istanbul ignore next */cov_1820g0n6e7.s[14]++;
        super.visitCompilationUnit(node);
        /* istanbul ignore next */cov_1820g0n6e7.s[15]++;
        this.indentLess();
        /* istanbul ignore next */cov_1820g0n6e7.s[16]++;
        return this.output;
    }
    visitFunctionDeclaration(node) {
        /* istanbul ignore next */cov_1820g0n6e7.f[5]++;
        cov_1820g0n6e7.s[17]++;

        this.write(\`FunctionDeclaration (name = \${node.name})\`);
        /* istanbul ignore next */cov_1820g0n6e7.s[18]++;
        this.indentMore();
        /* istanbul ignore next */cov_1820g0n6e7.s[19]++;
        super.visitFunctionDeclaration(node);
        /* istanbul ignore next */cov_1820g0n6e7.s[20]++;
        this.indentLess();
    }
    visitLiteralBoolean(node) {
        /* istanbul ignore next */cov_1820g0n6e7.f[6]++;
        cov_1820g0n6e7.s[21]++;

        this.write(\`LiteralBoolean: \${node.value}\`);
    }
    visitLiteralNumber(node) {
        /* istanbul ignore next */cov_1820g0n6e7.f[7]++;
        cov_1820g0n6e7.s[22]++;

        this.write(\`LiteralNumber: \${node.value}\`);
    }
    visitLiteralString(node) {
        /* istanbul ignore next */cov_1820g0n6e7.f[8]++;
        cov_1820g0n6e7.s[23]++;

        this.write(\`LiteralString: \${node.value}\`);
    }
    visitLiteralIdentifier(node) {
        /* istanbul ignore next */cov_1820g0n6e7.f[9]++;
        cov_1820g0n6e7.s[24]++;

        this.write(\`LiteralIdentifier: \${node.name}\`);
    }
    visitInvocationExpression(node) {
        /* istanbul ignore next */cov_1820g0n6e7.f[10]++;
        cov_1820g0n6e7.s[25]++;

        this.write(\`InvocationExpression:\`);
        /* istanbul ignore next */cov_1820g0n6e7.s[26]++;
        this.indentMore();
        /* istanbul ignore next */cov_1820g0n6e7.s[27]++;
        this.write(\`Target:\`);
        /* istanbul ignore next */cov_1820g0n6e7.s[28]++;
        this.indentMore();
        /* istanbul ignore next */cov_1820g0n6e7.s[29]++;
        node.target.visit(this);
        /* istanbul ignore next */cov_1820g0n6e7.s[30]++;
        this.indentLess();
        /* istanbul ignore next */cov_1820g0n6e7.s[31]++;
        this.write(\`Args:\`);
        /* istanbul ignore next */cov_1820g0n6e7.s[32]++;
        this.indentMore();
        /* istanbul ignore next */cov_1820g0n6e7.s[33]++;
        super.visitInvocationExpression(node);
        /* istanbul ignore next */cov_1820g0n6e7.s[34]++;
        this.indentLess();
    }
    indentMore() {
        /* istanbul ignore next */cov_1820g0n6e7.f[11]++;
        cov_1820g0n6e7.s[35]++;

        this.indent += this.indentBy;
    }
    indentLess() {
        /* istanbul ignore next */cov_1820g0n6e7.f[12]++;
        cov_1820g0n6e7.s[36]++;

        this.indent -= this.indentBy;
    }
    write(message) {
        /* istanbul ignore next */cov_1820g0n6e7.f[13]++;

        const indent = /* istanbul ignore next */(cov_1820g0n6e7.s[37]++, ' '.repeat(this.indent));
        /* istanbul ignore next */cov_1820g0n6e7.s[38]++;
        this.output += \`\${indent}\${message}\\n\`;
    }
} should emit a readable structure for a program 1`] = `
"CompilationUnit
  FunctionDeclaration (name = emptyFunction)
  FunctionDeclaration (name = returnsLiteral)
    LiteralBoolean: true
  FunctionDeclaration (name = hasExpressionBody)
    LiteralNumber: 1
    LiteralNumber: 2
    LiteralNumber: 3
  FunctionDeclaration (name = hasInvocationBody)
    InvocationExpression:
      Target:
        LiteralIdentifier: print
      Args:
        LiteralString: Hello World
"
`;
